### PR TITLE
[CDAP-17128] Add provisioning time metric

### PR DESCRIPTION
**Pull request by rmstar**

### CODE REVIEW

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramNotificationSubscriberService.java [hunk 1]
```diff
@@ -536,6 +536,37 @@ private void writeToHeartBeatTable(@Nullable RunRecordDetail recordedRunRecord,
         // Publish the program STARTING state before starting the program
         programStateWriter.start(programRunId, newProgramOptions, null, programDescriptor);
 
+        long provisioningTime = System.currentTimeMillis() / 1000 -
+          RunIds.getTime(programRunId.getRun(), TimeUnit.SECONDS);
+        // emit provisioning time metric
+        ProfileId profileId = null;
+        try {
+          RunRecordDetail runRecordMeta = appMetadataStore.getRun(programRunId);
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramNotificationSubscriberService.java]: The `recordProgramProvisioned` method call already returns the latest `RunRecordMeta` being recorded. You shouldn't need to read it again.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramNotificationSubscriberService.java [hunk 2]
```diff
@@ -536,6 +536,37 @@ private void writeToHeartBeatTable(@Nullable RunRecordDetail recordedRunRecord,
...
+          profileId = SystemArguments
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramNotificationSubscriberService.java]: In fact, the profileId, should already be in the `systemArgs` variable. You don't need the run record at all

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramNotificationSubscriberService.java [hunk 3]
```diff
@@ -536,6 +536,37 @@ private void writeToHeartBeatTable(@Nullable RunRecordDetail recordedRunRecord,
...
+            .getProfileIdFromArgs(programRunId.getNamespaceId(),
+              runRecordMeta.getSystemArgs()).orElse(null);
+        } catch (IOException e) {
+          LOG.error("Failed to get run record", e);
+        }
+
+        if (profileId != null) {
+          Map<String, String> args = programOptions.getArguments().asMap();
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramNotificationSubscriberService.java]: move this to a private function to have better separation of concern

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramNotificationSubscriberService.java [hunk 4]
```diff
@@ -536,6 +536,37 @@ private void writeToHeartBeatTable(@Nullable RunRecordDetail recordedRunRecord,
...
+            .put(Constants.Metrics.Tag.PROFILE, profileId.getProfile())
+            .put(Constants.Metrics.Tag.NAMESPACE, programRunId.getNamespace())
+            .put(Constants.Metrics.Tag.PROGRAM_TYPE, programRunId.getType().getPrettyName())
+            .put(Constants.Metrics.Tag.APP, programRunId.getApplication())
+            .put(Constants.Metrics.Tag.PROGRAM, programRunId.getProgram())
+            .put(Constants.Metrics.Tag.PROVISIONER, provisioner)
+            .put(Constants.Metrics.Tag.RUN_ID, programRunId.getRun())
+            .build();
+          metricsCollectionService.getContext(tags).gauge(Constants.Metrics.Program.PROGRAM_PROVISIONING_DELAY_SECONDS,
+            provisioningTime);
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramNotificationSubscriberService.java]: Misalign. Line break in argument should be aligned to the beginning of the method call:

```
metricsCollectionService.getContext(tags).gauge(Constants.Metrics.Program.PROGRAM_PROVISIONING_DELAY_SECONDS,
                                                provisioningTime);
```

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramNotificationSubscriberService.java (hunk 2 above)
Comment (rmstar) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramNotificationSubscriberService.java]: fixed.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramNotificationSubscriberService.java (hunk 3 above)
Comment (rmstar) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramNotificationSubscriberService.java]: done.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramNotificationSubscriberService.java (hunk 4 above)
Comment (rmstar) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramNotificationSubscriberService.java]: fixed.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramNotificationSubscriberService.java (hunk 1 above)
Comment (rmstar) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramNotificationSubscriberService.java]: removed, as it's not needed.

File: cdap-common/src/main/java/io/cdap/cdap/common/conf/Constants.java [hunk 5]
```diff
@@ -775,6 +776,7 @@
       public static final String PROGRAM_KILLED_RUNS = "program.killed.runs";
       public static final String PROGRAM_REJECTED_RUNS = "program.rejected.runs";
       public static final String PROGRAM_NODE_MINUTES = "program.node.minutes";
+      public static final String PROGRAM_PROVISIONING_DELAY_SECONDS = "program.provisioning.delay.seconds";
```
Comment (chtyim) [inline cdap-common/src/main/java/io/cdap/cdap/common/conf/Constants.java:779]: Is this only for unit test? If it is, define it in the test class

Comment (rmstar) [reply cdap-common/src/main/java/io/cdap/cdap/common/conf/Constants.java:779]: It's not only for tests. This metric will be read by the metrics writer extension and published to stackdriver.

**Review by chtyim — APPROVED** [review]

**Pull request merged**